Tna/feature documentation zkir v3 (#617)
midnightntwrk/midnight-ledger: spec/zkir.md

# ZKIR Instruction Set Reference

ZKIR ("zero-knowledge intermediate representation") is the low-level,
serialisable representation of a Midnight circuit.  The Compact compiler lowers

> each exported circuit to a ZKIR `IrSource`; the proving stack then turns that

Reviewer: The Compact compiler does not do this (it doesn't even know anything about IrSource).

ZKIR has two specified representations: a JSON one and a serialized binary one.  The Compact compiler generates ZKIR JSON outputs.

Author: Changed to the JSON representation since the doc shouldn't be learning on RUST implementation details that are irrelevant.

`IrSource` into a concrete PLONK circuit, generates prover / verifier keys, and
produces proofs.

ZKIR sits between Compact and the proof system:

```
Compact circuit  ──compile──▶  ZKIR (IrSource)  ──synthesise──▶  PLONK circuit  ──prove──▶  Proof
```

> This document specifies **IR v3** — the IR defined by the `zkir-v3` crate

Reviewer: No such thing as "IR v3".  Call this ZKIR version 3.0.

Author: Typo, fixed to be correct.

(`midnight-zkir-v3` 3.0.0-rc.2), which is the version integrated into Ledger 9
(consumed by `proof-server` and published to JS / TS as
`@midnight-ntwrk/zkir-v3`).  The previous stable IR, **v2** (`zkir` crate —
`midnight-zkir` 2.2.0), is retained for reference in
[Appendix B](#appendix-b--ir-v2-legacy); [§8](#8-migrating-from-v2) summarises
the v2 → v3 deltas.

> > A ZKIR program is **not** a stack machine and **not** a register machine in

Reviewer: This doesn't really make sense, but it's wrong anyway.

First of all, a program is not any kind of machine.

Second, ZKIR v2 can be thought of as both a language for a stack machine (which never pops and has random access instructions) and a language for a register machine (where the output register names are implicit in the position of the instructions).

ZKIR v3 is most definitely a language for a register machine.

Author: Re-written, can you please check? I do wonder if this detail is even needed.

> the usual sense.  It is a flat list of instructions over a **named memory**
> of typed values.  Each producing instruction reads its operands and binds
> its output(s) to fresh, explicitly-named variables; a variable is written
> once and never reassigned (SSA by name).  Operands are therefore either a
> variable reference (`%name`) or an inline immediate (`0x…`) — not positional
> tape indices as in v2.

> The same `IrSource` drives two passes:

Reviewer: I would not write this in terms of `IrSource`, that's a Rust implementation detail.  A better way is to describe this in terms of the JSON representation (like https://github.com/midnightntwrk/midnight-architecture/pull/131).

The Rust implementation happens to be a two-pass implementation, but that is not essential as far as I know.

Author: Reworked.

1. **Off-circuit preprocessing** (`IrSource::preprocess`, used by both `prove`
   and `check`) — concretely evaluates the program against a `ProofPreimage` to
   populate witness values and to derive the public inputs.  This is where
   most "this is undefined behaviour / this errors" rules are actually
   enforced at runtime.
2. **In-circuit synthesis** (`IrSource::circuit`) — emits the PLONK constraints
   that enforce each instruction, using the `midnight-zk` standard library
   (`ZkStdLib`).

## 1. The execution / circuit model

### 1.1 `IrSource` (v3)

A v3 circuit is described by the following structure:
* `version`: Major and minor IR version.
* `inputs`: The circuit's input variables, each a `(name, type)` pair. These are bound in memory before execution, decoded from `ProofPreimage::inputs`.
* `outputs`: The circuit's return signature - an explicit, positional list of result types. The `Output` terminator is type-checked against this.
* `do_communications_commitment`: Whether the circuit binds a commitment over its inputs and outputs ([§1.5](#15-communications-commitment)).
* `instructions`: The instruction list, executed in order.

> Note the change from v2's single `num_inputs: u32` to a fully-typed signature

Reviewer: This doesn't belong here.

Author: Removed completely.

**seven** variants. Each type is represented by a number of native field elements.
* `Scalar<BLS12-381>` — the BLS12-381 scalar field. This is the native type 
   for midnight-zk proofs and, therefore, it is represented as a native field element.
* `Point<Jubjub>` — A point of the embedded curve Jubjub, whose base field 
  is the BLS12-381 scalar field. Its raw encoding is two native elements (x, y), 
  representing its coordinates in affine form.
* `Scalar<Jubjub>` — Jubjub scalar field, used as the scalar argumenti to 
  `EcMul` / `EcMulGenerator`. Represented as a single native element.
* `Scalar<Secp256k1>` — SECP256k1 scalar field, emulated over the native
  field. Its raw encoding is two native elements.
* `Base<Secp256k1>` — SECP256k1 base field, emulated over the native
  field. Its raw encoding is two native elements.
* `Point<Secp256k1>` — A point from the SECP256k1 curve, emulated over the 
  native field. It is defined by the tuple `(x, y, is_id)`, where `x` and `y` are its
  coordinates (represented as `Base<Secp256k1>`) and `is_id` is a flag 
  determining whether the point is the identity or not. Its raw encoding is five 
  native elements. 

> * **`Bytes32`** — a 32-byte value. Its raw encoding is 32 native elements, 

Reviewer 2: I would be surprised if it is a whole 32 native elements, it would be very wasteful.

Author: You are correct, the raw encoding is 2 field elements, not 32. I've updated the docs.

#### `Bytes32FromLowHigh { inputs, output }`
* **Operands:** `inputs` — a pair `(low, high)`. **Outputs:** 1 — a `Bytes32`
  concatenating the first 31 bytes of `low` with byte `high`.
* **Precondition:** `low < 2^248` (fits in 31 bytes) and `high < 256` (one byte);
  off-circuit violation errors, in-circuit it is unsatisfiable.
* **Note:** the inverse of `Bytes32IntoLowHigh`; same temporary-bridge caveat.

#### `JubjubScalarFromNative { native, output }`
* **Operands:** `native` (`Native`). **Outputs:** 1 — a `JubjubScalar`.
* **Semantics:** casts a native value to a Jubjub scalar, reducing modulo the
  Jubjub scalar field order if necessary.
* **Note:** intended to be removed once a `BigUint` type is available.

### 3.7 Hashing

#### `TransientHash { inputs, output }`
* **Operands:** `inputs: Vec<Operand>` (all `Native`). **Outputs:** 1 — `H(inputs)`.
* **Semantics:** **Poseidon** hash (over the native scalar field) of the listed
  values. Circuit-friendly and cheap in-circuit; use for transient / ephemeral
  hashing.

#### `PersistentHash { alignment, inputs, outputs }`
* **Operands:** `alignment: Alignment`, `inputs: Vec<Operand>` (all `Native`).

>   **Outputs:** 1 — `Bytes32`.

Reviewer 3: PersistentHash requires exactly 2 outputs.

Reviewer: It's actually been changed in an unmerged PR (#629) that will form part of the final 3.0 ZKIR version.

But it raises a good point that we should very carefully audit this, or possibly add a disclaimer.  The 3.0 version is defined by what we release, not by this document --- so it's not a specification in that sense.

Author: Updated and yes, the audit should be off the code since the docs will fall out of step.

   enforced at runtime.
2. **In-circuit synthesis** (`IrSource::circuit`) — emits the PLONK constraints
   that enforce each instruction, using the `midnight-zk` standard library
   (`ZkStdLib`).

## 1. The execution / circuit model

### 1.1 `IrSource` (v3)

A v3 circuit is described by this structure (`zkir-v3/src/ir.rs`, serialisation
tag `ir-source[v3-generic]`):

| Field | Type | Meaning |
|---|---|---|
| `version` | `IrMinorVersion` | Minor IR version. v3 currently defines only `V0` (default). |
| `inputs` | `Vec<TypedIdentifier>` | The circuit's input variables, each a `(name, type)` pair. These are bound in memory before execution, decoded from `ProofPreimage::inputs`. |
| `outputs` | `Vec<IrType>` | The circuit's return signature — an explicit, positional list of result types. The `Output` terminator is type-checked against this. |
| `do_communications_commitment` | `bool` | Whether the circuit binds a commitment over its inputs and outputs ([§1.5](#15-communications-commitment)). |
| `instructions` | `Arc<Vec<Instruction>>` | The instruction list, executed in order. |

Note the change from v2's single `num_inputs: u32` to a fully-typed signature

> (`inputs` + `outputs`).

Reviewer 4: ```suggestion
A v3 circuit is described by the following structure:
* `version`: Major and minor IR version.
* `inputs`: The circuit's input variables, each a `(name, type)` pair. These are bound in memory before execution, decoded from `ProofPreimage::inputs`.
* `outputs`: The circuit's return signature - an explicit, positional list of result types. The `Output` terminator is type-checked against this.
* `do_communications_commitment`: Whether the circuit binds a commitment over its inputs and outputs ([§1.5](#15-communications-commitment)).
* `instructions`: The instruction list, executed in order.

Note the change from v2's single `num_inputs: u32` to a fully-typed signature
(`inputs` + `outputs`).
```

### 1.2 Named memory

* Memory is a map from **identifier** to **typed value** (`HashMap<Identifier, IrValue>`).
* It starts populated with the `inputs`, decoded from `ProofPreimage::inputs`.
  Each input consumes `IrType::encoded_len()` raw field elements (see

>   [§2.1](#21-types-irtype)); a shortfall errors with `Not enough raw inputs`,

Reviewer 4: ```suggestion
  Values are encoded as native field elements, with IrType::encoded_len()
  specifying the number of elements required for each type (see §2.1); a 
  shortfall errors with `Not enough raw inputs`,
```

  fixed and listed per-instruction below.

### 1.3 The fields and curves

* **Native field (`Fr`)** — the **BLS12-381 scalar field** (`outer::Scalar` in
  `transient-crypto`). A `Native` value is one native field element.
* **Embedded curve** — **Jubjub**, whose base field is the BLS12-381 scalar
  field (which is why Jubjub coordinates fit directly in native field
  elements). In v3 a curve point is a **first-class typed value**
  (`JubjubPoint`), not a raw coordinate pair; its raw encoding is two `Native`
  elements (x, y).
* **Embedded scalar field** — the Jubjub scalar field (`JubjubScalar`), used as
  the scalar argument to `EcMul` / `EcMulGenerator`.
* **Foreign curve** — **Secp256k1** (a.k.a. K256), emulated over the native
  field. Its point (`Secp256k1Point`), base field (`Secp256k1Base`), and scalar
  field (`Secp256k1Scalar`) are all first-class types, enabling
  Ethereum/Bitcoin-style ECDSA-adjacent circuits. These types pull in the
  `secp256k1` chip of the standard library only when used.
* **`Bytes32`** — a 32-byte value, used as the bridge between field elements and
  byte-oriented data (and as the digest type for the byte conversions).

> 

Reviewer 4: ```suggestion
```

(otherwise preprocessing fails with `Transcripts not fully consumed`):

| Stream | Consumed by | Notes |
|---|---|---|
| `public_transcript_outputs` | `PublicInput` | The next public output value witnessed into memory. |
| `private_transcript` | `PrivateInput` | The next private (witness-only) value into memory. |
| `public_transcript_inputs` | `Impact` | The declared public inputs; `Impact` reconciles them against this stream. |

`PublicInput` / `PrivateInput` consume `val_t.encoded_len()` raw elements per
(active) read and decode them into a typed value.

**Guards.** `PublicInput` and `PrivateInput` take an optional `guard` operand;
`Impact` takes a (required) `guard` operand. The guard is a boolean memory
value. When a guard is present and `false`, the instruction takes its
"inactive" path: `PublicInput` / `PrivateInput` bind the type's default value
without advancing their transcript cursor, and `Impact` declares `n` zeros (for
its `n` inputs) instead of the real values and records a skip. This is how
conditional (guarded) Impact operations are encoded. In-circuit, the guard does
**not** add constraints on `PublicInput` / `PrivateInput`; for `Impact` the

> guard is enforced as an in-circuit `select(guard, x, 0)` per input.

Reviewer 4: Not very clear the role of guards. 
@tkerber @miguel-ambrona , could you review this part?

Author: Section removed from documentation

above).

### 1.6 Minor versions

`IrMinorVersion` is `#[non_exhaustive]` and currently defines only `V0` (the
default), tag `ir-minor-version[v3]`. v3 starts fresh at `V0`; it does not
carry over v2's minor-version history (which has since grown to `V0`/`V1`/`V2`).
The top-level loader (`IrSource::load`) accepts only major version `3`, minor
`0`.

## 2. Operand and value conventions

### 2.1 Types (`IrType`)

Values are typed. `IrType` (`zkir-v3/src/ir_types.rs`, tag `ir-type[v1]`) has
**seven** variants. `encoded_len` is the number of raw `Native` (`Fr`) elements
the type lowers to (used for inputs, transcript reads, and `Encode`):

| `IrType` | Serialised name | `encoded_len` |
|---|---|---|
| `Native` | `Scalar<BLS12-381>` | 1 |
| `Bytes32` | `Bytes<32>` | 2 |
| `JubjubPoint` | `Point<Jubjub>` | 2 (x, y) |
| `JubjubScalar` | `Scalar<Jubjub>` | 1 |
| `Secp256k1Point` | `Point<Secp256k1>` | 5 |
| `Secp256k1Base` | `Base<Secp256k1>` | 2 |
| `Secp256k1Scalar` | `Scalar<Secp256k1>` | 2 |

A `Bytes32` is encoded as two field elements: the low 31 bytes and the high
(32nd) byte. The `Secp256k1*` lengths reflect the native-field limb

> decomposition used to emulate the foreign curve.

Reviewer 4: ```suggestion

Values are typed. `IrType` (`zkir-v3/src/ir_types.rs`, tag `ir-type[v1]`) has
**seven** variants. Each type is represented by a number of native field elements.
* `Scalar<BLS12-381>` — the BLS12-381 scalar field. This is the native type 
   for midnight-zk proofs and, therefore, it is represented as a native field element.
* `Point<Jubjub>` — A point of the embedded curve Jubjub, whose base field 
  is the BLS12-381 scalar field. Its raw encoding is two native elements (x, y), 
  representing its coordinates in affine form.
* `Scalar<Jubjub>` — Jubjub scalar field, used as the scalar argumenti to 
  `EcMul` / `EcMulGenerator`. Represented as a single native element.
* `Scalar<Secp256k1>` — SECP256k1 scalar field, emulated over the native
  field. Its raw encoding is two native elements.
* `Base<Secp256k1>` — SECP256k1 base field, emulated over the native
  field. Its raw encoding is two native elements.
* `Point<Secp256k1>` — A point from the SECP256k1 curve, emulated over the 
  native field. It is defined by the tuple `(x, y, is_id)`, where `x` and `y` are its
  coordinates (represented as `Base<Secp256k1>`) and `is_id` is a flag 
  determining whether the point is the identity or not. Its raw encoding is five 
  native elements. 
* **`Bytes32`** — a 32-byte value. Its raw encoding is 32 native elements, 
  range constrained to `[0, 256)`.
```

Examples below use this form.

## 3. Instruction reference (IR v3)

There are **33** v3 instructions.  Each entry lists operands, the number of
values it binds ("outputs"), semantics, the value types it supports, and
preconditions / undefined behaviour (UB).  "UB" means the instruction does not
itself constrain the precondition — the caller must guarantee it, or the proof /
preprocessing may be unsound or error.

Many arithmetic/equality/selection instructions are **type-polymorphic**: they
dispatch on the runtime type of their operands and error
(`Unsupported …: {ty} …`) on an unsupported type or a type mismatch between
operands.

### 3.1 Field / curve arithmetic

#### `Add { a, b, output }`
* **Operands:** `a`, `b`. **Outputs:** 1 — `a + b`.
* **Types:** `Native`, `JubjubPoint`, `Secp256k1Point`, `Secp256k1Base`,
  `Secp256k1Scalar`. For the two point types this is **point addition** (this
  subsumes v2's separate `EcAdd`). Both operands must share the same type.

#### `Mul { a, b, output }`
* **Operands:** `a`, `b`. **Outputs:** 1 — `a * b`.
* **Types:** `Native`, `Secp256k1Base`, `Secp256k1Scalar` (field multiplication;

>   scalar·point is `EcMul`, not `Mul`).

Reviewer 4: ```suggestion
* **Types:** `Native`, `Secp256k1Base`, `Secp256k1Scalar`. Field multiplication.
```

#### `Neg { a, output }`
* **Operands:** `a`. **Outputs:** 1 — `-a`.
* **Types:** `Native`, `JubjubPoint`, `Secp256k1Point`, `Secp256k1Base`,

>   `Secp256k1Scalar` (additive negation; point negation for the point types).

Reviewer 4: ```suggestion
  `Secp256k1Scalar`. Additive negation (point negation for the point types).
```

#### `Inv { a, output }`
* **Operands:** `a`. **Outputs:** 1 — `a⁻¹`.
* **Types:** `Native`, `Secp256k1Base`, `Secp256k1Scalar`.
* **Errors:** inverting zero (`cannot invert zero of type …`); in-circuit the
  zero case is unsatisfiable.

### 3.2 Boolean and selection

#### `Not { a, output }`
* **Operands:** `a` (boolean). **Outputs:** 1 — `!a`.

> * **Precondition:** `a ∈ {0,1}` (runtime-checked as boolean during preprocessing).

Reviewer 4: ```suggestion
* **Operands:** `a ∈ {0,1}`. **Outputs:** 1 — `!a`.
* **Errors:** if `a ∉ {0,1}`. In-circuit that case is unsatisfiable.
```

#### `TestEq { a, b, output }`
* **Operands:** `a`, `b`. **Outputs:** 1 — boolean `a == b` (a `Native` `0`/`1`).
* **Types:** `Native`, `JubjubPoint`, `Secp256k1Point`, `Secp256k1Base`,
  `Secp256k1Scalar`. (Compare with `ConstrainEq`, which *asserts* equality

>   rather than computing a boolean.)

Reviewer 4: ```suggestion
  `Secp256k1Scalar`. Different from `ConstrainEq`, which *asserts* equality
  rather than computing a boolean.
```

#### `CondSelect { bit, a, b, output }`
* **Operands:** `bit` (boolean), `a`, `b`. **Outputs:** 1.
* **Semantics:** binds `a` if `bit == 1`, else `b`.
* **Types:** `Native`, `JubjubPoint`, `Secp256k1Point`, `Secp256k1Base`,
  `Secp256k1Scalar`.

> * **UB:** `bit` must be `0` or `1`.

Reviewer 4: ```suggestion
* **Operands:** `bit ∈ {0,1}`, `a`, `b`. **Outputs:** 1.
* **Semantics:** binds `a` if `bit == 1`, else `b`.
* **Types:** `Native`, `JubjubPoint`, `Secp256k1Point`, `Secp256k1Base`,
  `Secp256k1Scalar`.
* **Errors:** if `bit ∉ {0,1}`. In-circuit that case is unsatisfiable.
```

### 3.3 Constraints and assertions (bind **no** outputs)

#### `Assert { cond }`
* **Operands:** `cond` (boolean). **Outputs:** none.
* **Semantics:** requires `cond == 1`. Fails preprocessing with
  `Failed direct assertion` otherwise; in-circuit, asserts non-zero.

> * **UB:** `cond` must be `0` or `1`.

Reviewer 4: ```suggestion
* **Operands:** `cond ∈ {0,1}`. **Outputs:** none.
* **Semantics:** asserts `cond == 1`.
* **Errors:** Fails preprocessing with `Failed direct assertion` if 
  `cond ∉ {0,1}`. In-circuit that case is unsatisfiable.
```

#### `ConstrainEq { a, b }`
* **Operands:** `a`, `b`. **Outputs:** none.
* **Semantics:** constrains `a == b`.
* **Types:** `Native`, `JubjubPoint`, `Secp256k1Point`, `Secp256k1Base`,
  `Secp256k1Scalar`.

#### `ConstrainToBoolean { val }`
* **Operands:** `val`. **Outputs:** none.
* **Semantics:** constrains `val ∈ {0,1}` (`Native`).

#### `ConstrainBits { val, bits }`
* **Operands:** `val`, `bits: u32`. **Outputs:** none.
* **Semantics:** constrains `val` to fit in `bits` bits (i.e. `val < 2^bits`).
* **Precondition:** `bits < FR_BITS` (255). A larger bound errors with

>   `Excessive bit bound`.

Reviewer 4: ```suggestion
* **Errors:** `bits >= FR_BITS`. A bound larger than FR_BITS errors 
  with `Excessive bit bound`.
```

### 3.4 Bit decomposition and comparison

#### `DivModPowerOfTwo { val, bits, outputs }`
* **Operands:** `val`, `bits: u32`. **Outputs:** 2 — declared in this order:
  1. `val >> bits` (the quotient / high bits)
  2. `val & ((1 << bits) - 1)` (the remainder / low `bits` bits)
* **Precondition:** exactly 2 output names (`DivModPowerOfTwo requires exactly 2
  outputs` otherwise); `bits ≤ FR_BYTES_STORED * 8` (248), else `Excessive bit

>   count`.

Reviewer 4: ```suggestion
* **Errors:** if `bits >= FR_BYTES_STORED * 8`, with `Excessive bit count`.
```

#### `ReconstituteField { divisor, modulus, bits, output }`
* **Operands:** `divisor`, `modulus`, `bits: u32`. **Outputs:**
  1 — `(divisor << bits) | modulus`.
* **Semantics:** inverse of `DivModPowerOfTwo`. Guarantees `modulus < 2^bits`
  and that the reconstituted value does not overflow the field (errors:
  `Reconstituted element overflows field`, or `Excessive bit count` if

>   `bits > 248`).

Reviewer 4: ```suggestion
  and that the reconstituted value does not overflow the field.
  * **Errors:** if if `bits > 248`, fails with `Reconstituted element overflows field`, 
    or `Excessive bit count`.
```

#### `LessThan { a, b, bits, output }`
* **Operands:** `a`, `b`, `bits: u32`. **Outputs:** 1 — boolean `a < b` (`Native`).
* **Semantics:** compares `a` and `b` as `bits`-bit unsigned integers.

> * **UB:** `a` and `b` must each fit in `bits` bits.

Reviewer 4: ```suggestion
* **Errors:** if `a` or `b` do not fit in `bits` bits. In-circuit that case is unsatisfiable.
```

  concatenating the first 31 bytes of `low` with byte `high`.
* **Precondition:** `low < 2^248` (fits in 31 bytes) and `high < 256` (one byte);
  off-circuit violation errors, in-circuit it is unsatisfiable.
* **Note:** the inverse of `Bytes32IntoLowHigh`; same temporary-bridge caveat.

#### `JubjubScalarFromNative { native, output }`
* **Operands:** `native` (`Native`). **Outputs:** 1 — a `JubjubScalar`.
* **Semantics:** casts a native value to a Jubjub scalar, reducing modulo the
  Jubjub scalar field order if necessary.
* **Note:** intended to be removed once a `BigUint` type is available.

### 3.7 Hashing

#### `TransientHash { inputs, output }`
* **Operands:** `inputs: Vec<Operand>` (all `Native`). **Outputs:** 1 — `H(inputs)`.
* **Semantics:** **Poseidon** hash (over the native scalar field) of the listed
  values. Circuit-friendly and cheap in-circuit; use for transient / ephemeral
  hashing.

#### `PersistentHash { alignment, inputs, outputs }`
* **Operands:** `alignment: Alignment`, `inputs: Vec<Operand>` (all `Native`).
  **Outputs:** 2 — the digest in the binary (field-aligned) format (exactly 2

>   output names, else `PersistentHash requires exactly 2 outputs`).

Reviewer 4: ```suggestion
  **Outputs:** 1 — `Bytes32`.
```

* **Semantics:** **SHA-256** hash. Inputs are first parsed according to
  `alignment` into a field-aligned-binary (FAB) byte representation
  ([field-aligned-binary.md](./field-aligned-binary.md)), then SHA-256 is taken
  over those bytes; the 32-byte digest is returned as 2 field elements (low 31
  bytes, then the high byte). Errors with `Inputs did not match alignment` if

>   `inputs` don't conform to `alignment`.

Reviewer 4: ```suggestion
  over those bytes. 
  * **Errors:** If `inputs` don't conform to `alignment`, errors with 
    `Inputs did not match alignment`.
```

* **Note:** much more expensive in-circuit than `TransientHash` (SHA-256 vs
  Poseidon); use where the digest must be a standard, stable hash across
  contexts (e.g. on-chain identifiers, interop).

#### `Keccak256 { alignment, inputs, outputs }`
* **Operands:** `alignment: Alignment`, `inputs: Vec<Operand>` (all `Native`).
  **Outputs:** 2 — the digest in the binary (field-aligned) format.
* **Semantics:** as `PersistentHash`, but computes the **Keccak-256** digest.
  Useful for EVM / Ethereum-compatible hashing. (`PersistentHash` and
  `Keccak256` share the same preprocessing / synthesis path, differing only in

>   the hash primitive.)

Reviewer 4: ```suggestion
* **Operands:** `alignment: Alignment`, `inputs: Vec<Operand>` (all `Native`).
  **Outputs:** 1 — `Bytes32`.
* **Semantics:** as `PersistentHash`, but computes the **Keccak-256** digest.
```

| `TestEq` | `a, b` | 1 (bool) |
| `CondSelect` | `bit` (bool), `a, b` | 1 |
| `Assert` | `cond` (bool) | 0 |
| `ConstrainEq` | `a, b` | 0 |
| `ConstrainToBoolean` | `val` | 0 |
| `ConstrainBits` | `val, bits` | 0 |
| `DivModPowerOfTwo` | `val, bits` | 2 |
| `ReconstituteField` | `divisor, modulus, bits` | 1 |
| `LessThan` | `a, b, bits` | 1 (bool) |
| `Copy` | `val` | 1 |
| `Encode` | `input` | `encoded_len` (1–5, per input type) |
| `IntoCoordinates` | `point` | 2 |
| `FromCoordinates` | `inputs` (x, y) | 1 (point) |
| `IntoBytes32` | `input` | 1 (Bytes32) |
| `FromBytes32` | `bytes, type` | 1 |
| `Bytes32IntoLowHigh` | `bytes` | 2 |
| `Bytes32FromLowHigh` | `inputs` (low, high) | 1 (Bytes32) |
| `JubjubScalarFromNative` | `native` | 1 (JubjubScalar) |
| `TransientHash` | `inputs[]` | 1 |
| `PersistentHash` | `alignment, inputs[]` | 2 |
| `Keccak256` | `alignment, inputs[]` | 2 |
| `EcMul` | `a, scalar` | 1 (point) |
| `EcMulGenerator` | `scalar` | 1 (point) |
| `HashToCurve` | `inputs[]` | 1 (point) |
| `PublicInput` | `guard?, type` | 1 |
| `PrivateInput` | `guard?, type` | 1 |
| `Impact` | `guard, inputs[]` | 0 (declares PIs) |
| `Output` | `vals[]` | 0 (terminator) |

> 

Reviewer 4: I believe we already have this above.

```suggestion
```
